torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4064,6 +4067,42 @@ Tensor linalg_det_backward(
   }
 }
 
+std::tuple<Tensor, Tensor, Tensor> attn_backward(
+    const Tensor& grad_o,
+    const Tensor& grad_a,
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v) {
+  if (!grad_o.defined() && !grad_a.defined()) {
+    return {};
+  }
+
+  Tensor a = at::tanh(at::matmul(q, k.transpose(0, 1)));
```

Comment:
Typically we don't want to have to recompute these expensive matmuls. Since you do return `a` as an output of `attn`, we can actually save that for backward and pass that as an input to this function.

How it works is whatever variables you reference, e.g. `result0` or `result1` the codegen sees that string and generates the code that saves the corresponding variable for backward.

Reply:
For this reason we can only save for backward when when we don't return a tensor for backward. If we still want to avoid recomputing it, it would be more tricky.